src/Enzo/enzo_EnzoReconstructorPLM.hpp

```diff
@@ -0,0 +1,406 @@
+   Grouping &primr_group, int dim, EnzoEquationOfState *eos, int stale_depth)
+{
+  std::vector<std::string> group_names = this->group_names_;
+
+  EnzoFieldArrayFactory array_factory(block, stale_depth);
+  EnzoPermutedCoordinates coord(dim);
+
+  Limiter limiter_func = Limiter();
+  const enzo_float theta_limiter = theta_limiter_;
+
+  // unecessary values are computed for inside faces of outermost ghost zone
+  for (unsigned int group_ind=0; group_ind<group_names.size(); group_ind++){
+
+    // load group name and number of fields in the group
+    std::string group_name = group_names[group_ind];
+    int num_fields = prim_group.size(group_name);
+
+    // Handle possibility of having a density floor
+    enzo_float prim_floor =0;
+    bool use_floor = false;
+    if (group_name == "density"){
+      prim_floor = eos->get_density_floor();
+      use_floor=true;
+    } else if (group_name == "pressure"){
+      prim_floor = eos->get_pressure_floor();
+      use_floor=true;
+    }
+
+    // iterate over the fields in the group
+    for (int field_ind=0; field_ind<num_fields; field_ind++){
+      // Cast the problem as reconstructing values at:
+      //   wl(k, j, i+3/2) and wr(k,j,i+1/2)
+
+      // Prepare cell-centered arrays
+      // define:   wc_left(k,j,i)   -> w(k,j,i)
+      //           wc_center(k,j,i) -> w(k,j,i+1)
+      //           wc_right(k,j,i)  -> w(k,j,i+2)
+      EFlt3DArray wc_left, wc_center, wc_right;
+      wc_left = array_factory.from_grouping(prim_group, group_name, field_ind);
+      wc_center = coord.left_edge_offset(wc_left, 0, 0, 1);
+      wc_right  = coord.left_edge_offset(wc_left, 0, 0, 2);
+
+      // Prepare face-centered arrays
+      // define:   wl_offset(k,j,i)-> wl(k,j,i+3/2)
+      //           wr(k,j,i)       -> wr(k,j,i+1/2)
+      EFlt3DArray wr, wl, wl_offset;
+      wr = array_factory.reconstructed_field(primr_group, group_name, field_ind,
+					     dim);
+      wl = array_factory.reconstructed_field(priml_group, group_name, field_ind,
+					     dim);
+      wl_offset = coord.left_edge_offset(wl, 0, 0, 1);
+      
+      // At the interfaces between the first and second cell (second-to-
+      // last and last cell), along a given axis, no need to worry about
+      // initializing the left (right) interface value thanks to the adoption
+      // of immediate_staling_rate
+
+      for (int iz=0; iz<wc_right.shape(0); iz++) {
+	for (int iy=0; iy<wc_right.shape(1); iy++) {
+	  for (int ix=0; ix<wc_right.shape(2); ix++) {
+
+	    // compute limited slopes
+	    enzo_float val = wc_center(iz,iy,ix);
+	    enzo_float dv = limiter_func(wc_left(iz,iy,ix), val,
+					 wc_right(iz,iy,ix), theta_limiter);
+	    enzo_float half_dv = dv*0.5;
+	    enzo_float left_val, right_val;
+
+	    if (use_floor) {
+	      right_val = EnzoEquationOfState::apply_floor(val - half_dv,
+							   prim_floor);
+	      left_val  = EnzoEquationOfState::apply_floor(val + half_dv,
+							   prim_floor);
```

Comment:
From a physical point of view, wouldn't it make more sense to actually limit the slope so that the reconstructed values are at most at the floor, rather than setting those value to the floor values?
Limiting the slopes would also ensure conservation of the reconstructed variable.

Reply:
This might make sense, but I really don't want to touch this before merging in. It seems like the exact sort of thing that we could break if we try to do this quickly before merging it in. I will happily open an issue on the topic.